### ISSUE TITLE
/tmp has wrong permissions on install

### ISSUE THREAD

**Issue by rardiol** [NixOS/nixpkgs#22413]
## Issue description

My /tmp was set to be root writable only after install, which broke a lot of things. I assume that's because I had created the tmp folder myself as root before running nixos-install while creating a btrfs subvolume hierarchy for the install. 

https://github.com/NixOS/nixpkgs/blob/56904d7c423f2b13b37fbd29f39bbb4b52bc7824/nixos/modules/installer/tools/nixos-install.sh#L93 apparently creates, but if's already there I think the permission doesn't change

### Steps to reproduce

Run the unstable usb installer.
Format a partition as btrfs, and create subvolumes, including /tmp as root.
Finish installing.
Try running things that use /tmp as user, for example lxqt's menu or pulseaudio.

## Technical details

* System: 17.03pre99759.f66d782 (Gorilla)  - same version installer

**Comment by grahamc** [comment]
Can you describe what you expect to happen, as well?

**Comment by rardiol** [comment]
I expect nixos-install to chmod /tmp if it's already there.

**Comment by lheckemann** [comment]
(not reproduced this issue, nor tried, but...) seems to me like having /tmp as a separate partition causes /tmp not to have mode 777 as it normally should?

**Comment by rardiol** [comment]
I haven't reinstalled to try it, but I don't think the problem is specifically if /tmp is a separate partition( and it's not in my case, it's a btrfs subvolume). I guess it always happens if /tmp is already there with the wrong permissions before running nixos-install, be it a subvolume, a directory or a partition.

**Comment by lheckemann** [comment]
Just tested in a VM, can confirm it happens when I create a separate partition for /tmp. Will test with just creating the dir.

**Comment by lheckemann** [comment]
Occurs if the directory is already present but not a separate mount as well. Apparently `mkdir -p` with `-m`, as used in nixos-install.sh, does not set the permissions if the directory already exists. I suppose nixos-install.sh should create and then chmod the directory rather than using `mkdir -m`?

**Comment by globin** [comment]
Might actually be fixed by chance in #22941 if switch-to-configuration is used in nixos-install.

**Comment by copumpkin** [comment]
I'm seeing this in images I'm building on a recent nixpkgs from a few days ago. Not sure if it's related, but I'm getting:

```
[root@hostname:~]# ls -ld /tmp/
drwxr-xr-t 12 root root 4096 Aug  8 20:52 /tmp/    
```

@globin could this be related to tmpfiles stuff repermissioning the `/tmp` on the image?
